core/rpc/json/openrpc/reflect.go

```diff
@@ -0,0 +1,238 @@
+	"reflect"
+	"slices"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/kwilteam/kwil-db/core/types"
+)
+
+type MethodDefinition struct {
+	Description  string
+	Summary      string
+	RequestType  reflect.Type
+	ResponseType reflect.Type
+	RespTypeDesc string
+	// Errors ...
+}
+
+// InventoryAPI iterates through all handler types, making a map of all unique
+// OpenRPC schemas and building a []Method with every Param populated and
+// referencing a Schema if an "object" instead of a plain type.
+func InventoryAPI(handlerDefs map[string]*MethodDefinition, knownSchemas map[reflect.Type]Schema) []Method {
+	if knownSchemas == nil {
+		knownSchemas = make(map[reflect.Type]Schema)
+	}
+
+	for _, def := range handlerDefs {
+		reflectTypeInfo(def.ResponseType, knownSchemas)
+		// Request types do not become known schemas since we treat each field
+		// of the request type as a parameter.
+	}
+
+	var methods []Method
+
+	for method, def := range handlerDefs {
+		// Request params from the request type
+		reflectTypeInfo(def.RequestType, knownSchemas)
+		reqSchema := knownSchemas[def.RequestType] // the non-$ref schema with Properties
+		if def.RequestType.Kind() == reflect.Struct {
+			delete(knownSchemas, def.RequestType)
+		}
+		params := make([]Param, 0, len(reqSchema.Properties))
+		for paramName, paramSchema := range reqSchema.Properties {
+			req := paramSchema.required
+			param := Param{
+				Name:     paramName,
+				Required: &req,
+				Schema:   paramSchema.Referenced(),
+				// Description: ? tag ?
+			}
+			params = append(params, param)
+		}
+		slices.SortFunc(params, func(a, b Param) int {
+			// Required params first, then alphabetical.
+			aReq := a.Required != nil && *a.Required
+			bReq := b.Required != nil && *b.Required
+			if aReq == bReq {
+				return cmp.Compare(a.Name, b.Name)
+			}
+			if aReq {
+				return -1
+			}
+			return 1
+		})
+
+		// Response param
+		respType := def.ResponseType
+		respSchema := knownSchemas[def.ResponseType] // reflectTypeInfo(respType, knownSchemas)
+		result := Param{
+			Name:        lowerFirstChar(respType.Name()),
+			Schema:      respSchema.Referenced(),
+			Description: def.RespTypeDesc,
+		}
+
+		methods = append(methods, Method{
+			Name:        method,
+			Params:      params,
+			Result:      result,
+			Description: def.Description,
+			// Summary: , ???
+			// Errors...
+		})
+	}
+
+	slices.SortFunc(methods, func(a, b Method) int {
+		return cmp.Compare(a.Name, b.Name)
+	})
+
+	return methods
+}
+
+func lowerFirstChar(s string) string {
+	r, sz := utf8.DecodeRuneInString(s)
+	if sz == 0 {
+		return s
+	}
+	if r == utf8.RuneError {
+		return s
+	}
+	return strings.ToLower(string(r)) + s[sz:]
+}
+
+func typeToSchemaType(t reflect.Type) string {
+	// Some special cases first.
+	switch t {
+	case reflect.TypeOf((*big.Int)(nil)), reflect.TypeFor[big.Int]():
+		return "string"
+	case reflect.TypeFor[types.HexBytes]():
+		return "string"
+	case reflect.TypeFor[[]byte]():
+		return "string"
+	}
```

Comment:
Should we also include our UUID, decimal, and uint256 types? Alternatively, we could just check for a fmt.Stringer, which applies to all of them.

Reply:
I guess there aren't any definitions where we return these data types, however I would imagine there might be for UUID (e.g. for vote store related things)

Reply:
Yeah, very well could be.  The `String` idea's not bad.  Will make some changes here.

Reply:
Now covered recent data types that have `MarshalJSON` methods that return a `[]byte` that represents a JSON string, such as uint256, uuid, and decimal.
But AFAICT there's no special significance of a `fmt.Stringer` in any circumstance with json marshalling, and we cannot identify how a field will actually marshal without knowing the implementation of the `MarshalJSON` method defined for that type (or the parent type that contains this field type).
Updated this section of the code with docs explaining why we have to list exceptions here.